tfx/experimental/distributed_inference/graphdef_experiments/subgraph_partitioning/graph_partition.py

```diff
@@ -0,0 +1,455 @@
+  """Partition all the graphs.
+
+  Here, we partition each graph given its graph_def and output names.'
+
+  Args:
+    op_to_graph_def: {a graph name: a GraphDef proto}
+    op_to_output_names: {a graph name: a list of output name}
+
+  Returns:
+    {graph name: a list of execution specs}
+  """
+  op_to_execution_specs = {}
+  for op in op_to_graph_def:
+    execution_specs = _partition_one_graph(op_to_graph_def[op],
+                                           op_to_output_names[op])
+    op_to_execution_specs[op] = execution_specs
+  return op_to_execution_specs
+
+
+def _partition_one_graph(graph_def, output_names):
+  """Partition a graph_def.
+
+  Args:
+    graph_def: a GraphDef proto
+    output_names: a list of output node names
+
+  Returns:
+    A list of execution spec.
+  """
+  graph = _get_graph(graph_def)
+  node_name_to_node_def = _get_node_name_to_node_def(graph_def)
+  node_name_to_input_names = _get_node_name_to_input_names(graph_def)
+
+  remote_op_relations = _get_remote_op_relations(graph_def,
+                                                 node_name_to_node_def,
+                                                 node_name_to_input_names)
+
+  execution_specs = _get_execution_specs(graph_def,
+                                         graph,
+                                         node_name_to_node_def,
+                                         node_name_to_input_names,
+                                         remote_op_relations,
+                                         output_names)
+  return execution_specs
+
+
+def _get_graph(graph_def):
+  temp = tf.Graph()
+  with temp.as_default():
+    tf.import_graph_def(graph_def)
+    return tf.compat.v1.get_default_graph()
+
+
+def _get_node_name_to_node_def(graph_def):
+  return {node.name: node for node in graph_def.node}
+
+
+def _get_node_name_to_input_names(graph_def):
+  return {node.name: list(node.input) for node in graph_def.node}
+
+
+def _get_remote_op_relations(graph_def,
+                             node_name_to_node_def,
+                             node_name_to_input_names):
+  """Get the execution dependencies between remote ops.
+
+  The remote op children must be executed before executing a remote op.
+
+  Args:
+    graph_def: a GraphDef proto
+    node_name_to_node_def: {node name: a NodeDef proto}
+    node_name_to_input_names: {node name: a list of node input names}
+
+  Returns:
+    {remote op: a list of remote op children}
+  """
+  remote_op_relations = {}
+
+  for node in graph_def.node:
+    if _is_remote_op(node):
+      remote_op_relations[node.name] = _get_remote_op_children(
+          node.name,
+          node_name_to_node_def,
+          node_name_to_input_names)
+
+  return remote_op_relations
+
+
+def _get_remote_op_children(remote_op_name,
+                            node_name_to_node_def,
+                            node_name_to_input_names):
+  """Find the remote op children for a remote op.
```

Comment:
It looks like you are finding *ancestors*, not *children*. Is that correct?

In any case, I think the algorithm can be implemented without this function; see my other comment below.

Reply:
Replaced "children" with "parents". Yes, the nodes I'm finding need to be executed before the given remote op.